### DISCUSSION TITLE
Failing to add third party script files from public folder using beforeInteractive in _document.js

### DISCUSSION THREAD

**Discussion by samiulllah** [vercel/next.js#37660] [Help] [answered]
I'm trying to load 3rd party script files from a public folder using the code given below, but when I load the page in the browser it doesn't have those script files attached to it. 

```
import { Html, Head, Main, NextScript } from 'next/document'
import Script from 'next/script'

export default function Document() {
  return (
    <Html>
      <Head/>
      <body>
        <Main />
        <NextScript />
            <Script src="lib/js/jquery-3.3.1.min.js" strategy="beforeInteractive"/>
            <Script src="lib/js/bootstrap.min.js" strategy="beforeInteractive"/>
            <Script src="lib/js/jquery.nice-select.min.js" strategy="beforeInteractive"/>
            <Script src="lib/js/jquery-ui.min.js" strategy="beforeInteractive"/>
            <Script src="lib/js/jquery.slicknav.js" strategy="beforeInteractive"/>
            <Script src="lib/js/mixitup.min.js" strategy="beforeInteractive"/>
            <Script src="lib/js/owl.carousel.min.js" strategy="beforeInteractive"/>
            <Script src="lib/js/main.js" strategy="beforeInteractive" />
      </body>
    </Html>
  );
}
```

I've also asked this question on StackOverflow but no one was able to solve it <a href="https://stackoverflow.com/questions/72599597/what-is-proper-way-to-import-script-files-in-next-js">there.</a>

**Comment by icyJoseph** [comment, marked as answer]
Hi,

There's a bug on 12.1.6, that makes is to so that this is not working at the moment. Updating with more info shortly.

The last stable release 12.1.6 was back on May 2nd, and a lot has been happening.

More relevant links:

- https://github.com/vercel/next.js/pull/37000
- https://github.com/vercel/next.js/issues/37225#issuecomment-1142390675
- https://github.com/vercel/next.js/pull/37084,

### While the fix is published

```jsx
<script
  type="text/javascript"
  src="https://unpkg.com/lodash@4.17.20/lodash.js"
></script>
```

I'd say it is best to wait until a new release happens, and use a simple script for now.

**Reply by samiulllah** [reply]
> Hi,
> 
> There's a bug on 12.1.6, that makes is to so that this is not working at the moment. Updating with more info shortly.
> 
> The last stable release 12.1.6 was back on May 2nd, and a lot has been happening.
> 
> More relevant links:
> 
> * [Fixes `beforeInteractive` scripts failing in custom document #37000](https://github.com/vercel/next.js/pull/37000)
> * [Next/script inside _document.js won't load #37225 (comment)](https://github.com/vercel/next.js/issues/37225#issuecomment-1142390675)
> * [Update script.md #37084](https://github.com/vercel/next.js/pull/37084),
> 
> ### While the fix is published
> ```js
> <script
>   type="text/javascript"
>   src="https://unpkg.com/lodash@4.17.20/lodash.js"
> ></script>
> ```
> 
> I'd say it is best to wait until a new release happens, and use a simple script for now.

I've tried loading with simple script tag in _document.js and also tried in _app.js. It loads with simple script tag but it says hydration failed.

**Reply by icyJoseph** [reply]
Did you try placing them inside Head?

**Reply by samiulllah** [reply]
Yes inside head.

**Reply by icyJoseph** [reply]
The following works without problem. I see the code inside my script being executed.

Perhaps you are now seeing a different problem? Could you share a repository? Hydration failed could be because of divergent HTML between server and client.

```tsx
// Works
import { Html, Head, Main, NextScript } from "next/document";

export default function Document() {
  return (
    <Html>
      <Head>
        {/* eslint-disable-next-line @next/next/no-sync-scripts */}
        <script type="text/javascript" src="/lib/main.js" />
      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  );
}
```

**Reply by samiulllah** [reply]
@icyJoseph 

It's **loading scripts** but I think those loaded **scripts** aren't **executing**, all actions to be performed by them aren't working e.g menu isn't expanding & collapsing which is provided by **/lib/js/bootstrap.min.js**, although I checked in the browser this script is loaded. And that hydration error is occurring due to the code below,

```
   <nav className="humberger__menu__nav mobile-menu">
            <ul>
                <li className="active"><a href="./index.html">Home</a></li>
                <li><a href="./shop-grid.html">Shop</a></li>
                <li><a href="#">Pages</a>
                    <ul className="header__menu__dropdown">
                        <li><a href="./shop-details.html">Shop Details</a></li>
                        <li><a href="./shoping-cart.html">Shoping Cart</a></li>
                        <li><a href="./checkout.html">Check Out</a></li>
                        <li><a href="./blog-details.html">Blog Details</a></li>
                    </ul>
                </li>
                <li><a href="./blog.html">Blog</a></li>
                <li><a href="./contact.html">Contact</a></li>
            </ul>
        </nav>
       <div id="mobile-menu-wrap"></div>
```

When I commented  this code the hydration error disappear. Also please note I'm trying to use a template of bootstrap which works when not imported in next js. I replaced all class="" with className="" and imported all required files.

**Reply by icyJoseph** [reply]
Yeah, this is why it is usually a bad idea to mix frameworks, I guess here there's jQuery and Bootstrap's own JS, which are colliding with React's DOM manipulation.

Could you share a repository? I guess you are now facing divergent UI, which happens when the HTML rendered on the server is either invalid, or doesn't match what React encounters on the client. This is not a Next.js, nor React issue, but a pitfall of switching to SSR where one has to account for the server rendered HTML and the client first frame.

A repository to look at would help us solve your issue much faster. Could you put this into a small reproduction repo?

**Reply by samiulllah** [reply]
Alright. Please check my repo here ---> https://github.com/samiulllah/pillow_next_demo.git

**Reply by samiulllah** [reply]
As you were saying might be jquery and bootstrap js may be messing with React. But I've tested in a new project with bootstrap & Jquery everything works fine.

**Reply by icyJoseph** [reply]
I see no error... on this repository. Besides linting issues when running build, both `dev` and `prod` mode work without issue.

**Reply by samiulllah** [reply]
> I see no error... on this repository. Besides linting issues when running build, both `dev` and `prod` modes work without issue.

I've commented on lines causing hydration errors in index.js. Beside hydration issue next is loading scripts but not executing them because if you try to expand nav bar it won't work.

**Reply by samiulllah** [reply]
@icyJoseph  you can test it by creating the next app and adding **bootstrap nav** to it. You will come to know that **nav** won't **expand** or **collapse** because bootstrap CSS is loading and executing but its **js** is loaded but **not executing**. I just did this in a separate project too.

**Reply by icyJoseph** [reply]
Yeah, I see it in your repo now, I've dealt with this before, and it is not very easy, mostly because we are forcing two frameworks to work together. 

~~Let me take a look for some time, but overall~~ I'd say, you might want to consider `reactstrap` or `react-bootstrap`, here you are really setting up yourself to have problems. I am a bit worried because I see people do this more than I'd like to.

Anyway, one way to solve this, is to let the SSR and React do their thing, and then kick in with your jQuery DOM modifying code, though as I said before, you should really avoid this.

1. Replace the IIFE in `mainW.js` for:

```js
// of course you perhaps want better names
function _runJQueryControls($) {
// rest of function
}

window.runJQueryControls = () => _runJQueryControls(jQuery);
```

2. Inside `index.js` 

```js
 useEffect(() => {
    window?.runJQueryControls();
  }, []);
```

3. In `next.config.js`

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: false, 
};
```

4. It should work (just reload the page in case it complains)

As you can see, putting two frameworks to control the DOM is seed for trouble. In the initial days of React, what people used to do was let React control a small section of the DOM and then let whatever framework control the rest, but that doesn't really translate well to SSR days.

**Reply by samiulllah** [reply]
@icyJoseph followed your steps problem is solved. So the problem was **react** DOM and **Jquery** Dom operations were colliding?
And you solved it by first letting react/next to manipulate DOM then giving jquery to do its job in useEffect hook? Please correct me if I got it wrong. Thanks alot for your time.

**Reply by icyJoseph** [reply]
Yes exactly, the thing is that the server renders what people call dehydrated HTML, a simple string, but then on the client, as soon as the HTML arrives, then jQuery starts to modify the DOM as react is doing something called hydration (it renders the HTML on the client, expecting all nodes inside its root, `__next` in this case, to match what the server sends, in order to properly attach event listeners and give interactivity to the page), and this causes React to run into a problem, how to hydrate a DOM that no longer matches what it expects.

**Reply by samiulllah** [reply]
Thank you very much. I got it!

**Reply by icyJoseph** [reply]
Don't forget to mark as answered. This was a nice trip down the rabbit hole!

**Reply by colevr1** [reply]
Hi, I am struggling with an old website from a customer, and I found this thread very useful. Just one question, if I use the Next Link component, when coming back to the page, seems not working anymore, even if I was able to reload the useEffect by detecting router changes. Any clue on how to solve this without using the <a> tag? (which works of course). 
Thanks.
